model/task_lifecycle.go

```diff
@@ -31,111 +31,123 @@ type StatusChanges struct {
+		buildToTaskMap[t.BuildId] = t
+		if active {
+			// if the task is being activated, and it doesn't override its dependencies
+			// activate the task's dependencies as well
+			if !t.OverrideDependencies {
+				deps, err := task.GetRecursiveDependenciesUp(originalTasks, nil)
+				if err != nil {
+					catcher.Add(errors.Wrapf(err, "error getting tasks '%s' depends on", t.Id))
+				}
+				if t.IsPartOfSingleHostTaskGroup() {
+					for _, dep := range deps {
+						// reset any already finished tasks in the same task group
+						if dep.TaskGroup == t.TaskGroup && t.TaskGroup != "" && dep.IsFinished() {
+							if err = resetTask(dep.Id, caller, false); err != nil {
+								catcher.Add(errors.Wrapf(err, "error resetting dependency '%s'", dep.Id))
+							}
+						} else {
+							tasksToActivate = append(tasksToActivate, dep)
 						}
-					} else {
-						tasksToActivate = append(tasksToActivate, dep)
 					}
+				} else {
+					tasksToActivate = append(tasksToActivate, deps...)
+				}
+			}
+
+			// Investigating strange dispatch state as part of EVG-13144
+			if !utility.IsZeroTime(t.DispatchTime) && t.Status == evergreen.TaskUndispatched {
+				if err := resetTask(t.Id, caller, false); err != nil {
+					catcher.Add(errors.Wrap(err, "error resetting task"))
 				}
 			} else {
-				tasksToActivate = append(tasksToActivate, deps...)
+				tasksToActivate = append(tasksToActivate, originalTasks...)
 			}
-		}
 
-		// Investigating strange dispatch state as part of EVG-13144
-		if !utility.IsZeroTime(t.DispatchTime) && t.Status == evergreen.TaskUndispatched {
-			if err := resetTask(t.Id, caller, false); err != nil {
-				return errors.Wrap(err, "error resetting task")
+			if t.DistroId == "" && !t.DisplayOnly {
+				grip.Critical(message.Fields{
+					"message": "task is missing distro id",
+					"task_id": t.Id,
+				})
+			}
+
+			// If the task was not activated by step back, and either the caller is not evergreen
+			// or the task was originally activated by evergreen, deactivate the task
+		} else if !evergreen.IsSystemActivator(caller) || evergreen.IsSystemActivator(t.ActivatedBy) {
+			var err error
+			// deactivate later tasks in the group as well, since they won't succeed without this one
+			if t.IsPartOfSingleHostTaskGroup() {
+				tasksInGroup, err := task.FindTaskGroupFromBuild(t.BuildId, t.TaskGroup)
+				if err != nil {
+					catcher.Add(errors.Wrapf(err, "error finding task group '%s'", t.TaskGroup))
+				}
+				for _, taskInGroup := range tasksInGroup {
+					if taskInGroup.TaskGroupOrder > t.TaskGroupOrder {
+						originalTasks = append(originalTasks, taskInGroup)
+					}
+				}
+			}
+			if t.Requester == evergreen.MergeTestRequester {
+				if err = DequeueAndRestartForTask(nil, &t, message.GithubStateError, caller, fmt.Sprintf("deactivated by '%s'", caller)); err != nil {
+					catcher.Add(err)
+				}
 			}
-		} else {
 			tasksToActivate = append(tasksToActivate, originalTasks...)
+		} else {
+			return nil
```

Comment:
Shouldn't we continue to the next item in the loop?

Reply:
I had assumed that the ActivatedBy field would not change for any of the tasks in the list, and since the caller / active parameters do not change it'd be best to return early here (same setup as before this implementation).

Reply:
But the "else if" is related specifically to the task ActivatedBy, so it's possible that the first task would skip the else if because it wasn't activated by the right person, but the second task was.

Reply:
Ah, I had thought the tasks in the submitted batch would all have the same ActivatedBy and that wouldn't be a case to consider.  Ok, changing to continue in this case.